### ISSUE TITLE
Save seed (and generation parameters) in the world file

### ISSUE THREAD

**Issue by ftomassetti** [Mindwerks/worldengine#30]
"There seems to be no easy way to check the seeds that are used to generate the map. A little thing, but very frustrating when I want to recreate a map whose seed I did not write down. Perhaps the parameters used to create the world could be embedded/commented into the images as they are created? Or, perhaps, could this information be extracted from the .world file? "

We should save all generation parameters and also the Lands version number

**Comment by rbb** [comment]
In progress. I should have a commit in a day or two.

**Comment by ftomassetti** [comment]
Great to read that!

**Comment by rbb** [comment]
Hey Federico,

I'm working through the issues with getting the protobuf stuff working with the new fields/members in the world class. Adding members obviously changes the serialization schema, and therefore the world file format. Do you care about backward compatibility?

Russ

**Comment by ftomassetti** [comment]
Hey Russ,
I think we could add the new fields marking them as optional and assigning to them the next free field id, it should give us backward compatibility (of course we should use some sensible default value).
So said I think backward compatibility it is not super important because we just added protobuf, so I guess we can change it if it is necessary

**Comment by rbb** [comment]
Sounds reasonable. There might have to be a try:, except: clause to handle to the two versions. In my initial experiments, things seemed a little wonky.

What was the main motivation for adding protobuf? Giving access to the world data to other programming languages? Or does it have something to do with the GUI development?

On a side note, I've had a little bit of a setback. I'm having trouble getting protobuf installed via macports. I think when I recently reinstalled macports after upgrading to yosemite, that it screwed some things up, so I'm in the middle of re-installing macports. Normally this wouldn't be a big deal, but I make heavy use of the octave program, and it takes most of a day for macports to (re)compile it.

**Comment by rbb** [comment]
I started working on merging the master into my changes, and the new simulations/ directory was a little bit of a surprise. I'm going to back off on including seed, etc into the world class for now, and focus on the help message and verbose messages.

**Comment by rbb** [comment]
Hmm. simulations/ was always there. Grr. Still wrapping my head around git (still used to subversion).

**Comment by ftomassetti** [comment]
"Sounds reasonable. There might have to be a try:, except: clause to handle to the two versions. In my initial experiments, things seemed a little wonky."

Little wonky is ok, first have it working, we can improve it later.

The idea about protobuf was to support other languages (would be nice to be able to import lands generated worlds in programs written in Java, C/C++ etc.) but I was also hoping it was more space-efficient that pickle. In practice I think it is not :)

I am sorry about the problems with protobuf. I used brew on the mac and I ran ```brew install protobug``` however you need the protoc tool that comes with protobuf only to generate the stubs (running generated_protobuf_stubs.sh). To actually run lands after that you just need the protobuf python package installable through pip. If it becomes messy to install protobuf you can always send me the changes you have done to the proto file and I can generate the new stubs and send them back to you. It is best to use brain cycles on coding than on figuring out installation stuff :D

About the problems with git... I hear you, it is starting to become a little more natural now but I am still confused from time to time. If I can help do not hesitate to ask, if you send a pull request which require some adjusting we can work on that together and have it merged.

**Comment by ftomassetti** [comment]
Btw, I did some commits to master (moving code from geo to separate classes, one for each simulation) and it could have caused conflicts with your changes. It is not a super easy scenario because you have to play with two remotes (your github repo, my github repo) so it is something that can easily confuse.

**Comment by rbb** [comment]
whew. I think I finally got the git commit figured out  - including your commits to master over the weekend. ([This commit])(https://github.com/rbb/lands/tree/world_class_with_creation_params) still breaks the protobuf stuff, but at least it has the seed, and a few other variables included in the world class.

**Comment by ftomassetti** [comment]
Thank you! I took your branch, I am doing a few changes and fixing the protobuf serialization.
I just removed the verbose and the recursion limit fields from world (I think they are parameters of the generation important only while generating, but they should not be stored).

I will merge soon your changes in the master branch and add you in the README.md, if it is ok for you

**Comment by rbb** [comment]
Awesome.

I have a couple of thoughts about the simulations package/directory. Should these functions really be part of the World class? It seems like they really want to be.

Russ

**Comment by ftomassetti** [comment]
well, I think about the simulations as transformations operated on the world class. The world class should represent a specific state at a given time.
Each simulation could know for example when it can be applied, generate verbose messages (which could either go to the console or to a label in a GUI).
Ideally I would like to have a common interface for all the operations, so that I can treat them in the same way from the GUI. I introduced them when working on the GUI for this reason mainly.
I would also love to reduce the size of the World class.
However I think that the whole application could use some refactoring because it was not well planned at the beginning (I started writing it while learning Python and then it just sort of grow over time).
Now that we have more contributions we should have a cleaner, well architected (and documented code) so I am very open to suggestions. I will close this issue because we solved it with your code but feel free to open other issues or write to me.
Or maybe I will start a mailing-list in the future.
